### ISSUE TITLE
Copying reducers

### ISSUE THREAD

**Issue by opensdh** [kokkos/kokkos#6602]
[`parallel_reduce` with a reducer](https:/kokkos/kokkos/blob/1a3ea28f6e97b4c9dd2c8ceed53ad58ed5f94dfe/core/src/Kokkos_Parallel_Reduce.hpp#L1750) accumulates into a copy of the reducer, which breaks if the reducer contains the result value rather than a reference (or pointer) to it.  The comment above says "take by copy to allow for inline construction", but it's not clear to me what "inline construction" is.

Is there in fact a good reason to make the copy (for the reducer case)?  If so, should it be called out in the documentation?

**Comment by masterleinad** [comment]
The interface for reducers only has `reference()` and `view()`. Any value contained in a reducer would only be copied to the `GPU` but not back. To make that work we must only store pointers/references (possibly wrapped in a `View` or so).

**Comment by opensdh** [comment]
> The interface for reducers only has `reference()` and `view()`. Any value contained in a reducer would only be copied to the `GPU` but not back.

I'm not saying it _does_ this, but Kokkos could certainly use `Reducer::init` on the device to create a _new_ reduction accumulator and then copy the result back to the caller's object via its `reference()`.

> To make that work we must only store pointers/references (possibly wrapped in a `View` or so).

If the reducer does have a pointer, how does that help?  Now copying the reducer object onto the device produces an invalid pointer (or a `View` of the wrong memory space).

Sorry if I'm just confused; the "Question" label is well deserved.

**Comment by masterleinad** [comment]
> I'm not saying it _does_ this, but Kokkos could certainly use `Reducer::init` on the device to create a _new_ reduction accumulator and then copy the result back to the caller's object via its `reference()`.

Yes, we could do something like that (depending on the requirements for reducers to be device-constructible).

> If the reducer does have a pointer, how does that help? Now copying the reducer object onto the device produces an invalid pointer (or a `View` of the wrong memory space).

You still need to make sure that the `View` has the correct memory space. If it's device-accessible, we write to it on the device and don't copy the result back to the host. Note that (mem-) copying a pointer or a `View` to the device, gives you an new object with the same content and if the pointer referenced GPU memory it continues to do so and is usable. A reference however would always point to a host object and thus can't be used on the GPU.

The mental model is that all objects used in a kernel will be memcopied to the device but not back. I'm not sure if there is a good reason why we copy the reducer in the line you pointed to in your initial comment but this is why you can't use values in a reducer (or any other object used in a kernel) that you expect to be updated on the device.

**Comment by opensdh** [comment]
I'm afraid I still don't see what the intended usage pattern is: `memcpy`ing the [`SumMyArray` example reducer](https://kokkos.github.io/kokkos-core-wiki/ProgrammingGuide/Custom-Reductions-Custom-Reducers.html#example) would leave it with a pointer to host memory.  Does that example just not work with a GPU backend?

Would it be a better idea to try to figure it out from the source for the built-in reducers (and then propose a patch for the documentation)?

**Comment by masterleinad** [comment]
> I'm afraid I still don't see what the intended usage pattern is: `memcpy`ing the [`SumMyArray` example reducer](https://kokkos.github.io/kokkos-core-wiki/ProgrammingGuide/Custom-Reductions-Custom-Reducers.html#example) would leave it with a pointer to host memory. Does that example just not work with a GPU backend?

It's just important what memory space the `View` returned from the `view()` member function returns. If it's a host `View`, we use a different allocation on the device and (deep) copy it to that `View` after the kernel ran.

> Would it be a better idea to try to figure it out from the source for the built-in reducers (and then propose a patch for the documentation)?

Yes, it might help to look at https://github.com/kokkos/kokkos/blob/71a9bcae52543bd065522bf3e41b5bfa467d8015/core/src/Kokkos_Parallel_Reduce.hpp#L33-L72.

**Comment by opensdh** [comment]
I actually mostly read the implementation of `parallel_reduce`, and what I found is that `reference` (from `ReducerConcept`) isn't used (or even trait-checked) at all.  (I also see that `init` is called on objects that are out of lifetime, so it should really be doing placement-new on them, although that doesn't matter much if no one is calling destructors.)

**Comment by ajpowelsnl** [comment]
Hi @opensdh - can you suggest next steps (code and / or documentation ) for better-behaved reducers?

**Comment by opensdh** [comment]
I'd start by removing the mentions of `reference` in the documentation, since that name doesn't actually mean anything to Kokkos here and its presence is misleading.  Next, I'd recommend documenting that the reducer is copied (or is it `memcpy`ed?) to the device, but that its `view()` (which is, I think, called only on the host?) can still be a host view which will be populated from a temporary on the device.  (In general, which functions might be called on the host?  Which might be called on the device?  Having `KOKKOS_INLINE_FUNCTION` in the example _only_ where necessary and not on things like the constructor would be helpful.)

The existing documentation is presumably trying to convey (some of) this with phrases like "defined in the memory space where the object is used", but it's a bit opaque because it's not clear what "object" _or_ "used" mean there.
